tests/test_make_conf.py

```diff
@@ -0,0 +1,131 @@
+
+import attr
+import pytest
+
+from wgbs_pipeline.make_conf import get_parser, main, make_conf, write_conf
+
+
+@attr.s(auto_attribs=True)
+class StubArgs:
+    reference: str
+    extra_reference: str
+    underconversion_sequence: Optional[str] = None
+    include_file: Optional[str] = None
+    outfile: str = "gembs.conf"
+    num_threads: int = 8
+    num_jobs: int = 3
+
+
+@pytest.mark.parametrize(
+    "args,condition",
+    [
+        (
+            [
+                "-t",
+                "3",
+                "-j",
+                "8",
+                "-r",
+                "ref.fa.gz",
+                "-e",
+                "extra_ref.fa.gz",
+                "-u",
+                "chrL",
+                "-i",
+                "include.conf",
+                "-o",
+                "my.conf",
+            ],
+            does_not_raise(),
+        ),
+        (["-r", "ref.fa.gz", "-e", "extra_ref.fa.gz"], does_not_raise()),
+        (["-r", "ref.fa.gz"], pytest.raises(SystemExit)),
+        (["-e", "extra.fa.gz"], pytest.raises(SystemExit)),
+    ],
+)
+def test_parser(args: List[str], condition):
+    parser = get_parser()
+    with condition:
+        parser.parse_args(args)
+
+
+@pytest.mark.parametrize(
+    "args,expected_num_rows,assertions",
+    [
+        (
+            StubArgs("/foo/reference.fa.gz", "/bar/exref.fa.gz"),
+            11,
+            [
+                (0, "reference = reference/reference.fa.gz"),
+                (1, "extra_references = reference/exref.fa.gz"),
+                (4, "sequence_dir = ${base}/fastq/@SAMPLE"),
+                (9, "threads = 8"),
+                (10, "jobs = 3"),
+            ],
+        ),
+        (
+            StubArgs("reference.fa.gz", "exref.fa.gz", underconversion_sequence="chrL"),
+            13,
+            [(11, "[mapping]"), (12, "underconversion_sequence = chrL")],
+        ),
+        (
+            StubArgs("reference.fa.gz", "exref.fa.gz", include_file="my_dir/include.conf"),
+            13,
+            [(11, "[mapping]"), (12, "include my_dir/include.conf")],
+        ),
+        (
+            StubArgs(
+                "/foo/reference.fa.gz", "/bar/exref.fa.gz", num_threads=1, num_jobs=1
+            ),
+            11,
+            [(9, "threads = 1"), (10, "jobs = 1")],
+        ),
+        (
+            StubArgs(
+                "reference.fa.gz",
+                "exref.fa.gz",
+                underconversion_sequence="chrL",
+                include_file="include.conf",
+            ),
+            14,
+            [
+                (11, "[mapping]"),
+                (12, "underconversion_sequence = chrL"),
+                (13, "include include.conf"),
+            ],
+        ),
+    ],
+)
+def test_make_conf(args, expected_num_rows: int, assertions: List[Tuple[int, str]]):
+    output = make_conf(args)
+    assert len(output) == expected_num_rows
+    for row_index, expected_value in assertions:
```

Comment:
It's unlikely. There are a few edge cases in `make_conf()` that are only tested here. Particularly here, I want to make sure that I test it when one of, none of, and both of `args.underconversion_sequence` and `args.include_file` are `None`. I just didn't want to assert the whole list is equal, since it is pretty redundant (most items in the list are constant).